### PR TITLE
HTCONDOR-779 condor_adstash from file

### PR DESCRIPTION
**Pull request by jasoncpatton**
Allows `condor_adstash` users to specify `--ads ADFILE` to parse and push history from a file containing job history classads to Elasticsearch.

# HTCondor Pull Request Checklist for internal reviewers

- [x] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [x] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [x] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed
- [ ] Check for version history, if needed
- [ ] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [ ] Check that commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

**Review by markcoatsworth — CHANGES_REQUESTED** [review]
Code mostly looks good. I unfortunately wasn't able to test it locally because Jason indicated I would need an Elasticsearch server and code freeze is tonight. I've pointed out a couple of issues that either need to be fixed or explained.

File: src/condor_scripts/adstash/history.py [hunk 1]
```diff
@@ -32,6 +32,80 @@
...
+    """
+    Given an iterator of ads, process its entire set of history
+    """
+    logging.info(f"Start processing the adfile: {job_ad_file}")
+
+    my_start = time.time()
+    buffered_ads = {}
+    count = 0
+    total_upload = 0
+    if not args.read_only:
```
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_scripts/adstash/history.py:45]: Yes, `--read_only` doesn't work exactly as intended, so it's only useful for testing part of the code, this is something that needs to be fixed in a later release.

[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_scripts/adstash/history.py:45]: For now, leaving it to have the same behavior as the other query methods.

File: docs/man-pages/condor_adstash.rst [hunk 2]
```diff
@@ -11,7 +11,7 @@ Synopsis
 
 **condor_adstash** [**--help** ]
 
-**condor_adstash** [**--process_name** *NAME*] [**--standalone** ]
+**condor_adstash** [**--process_name** *NAME*] [**--standalone** ] [**--ads** *ADFILE*]
```
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply docs/man-pages/condor_adstash.rst]: Changed to `--ad_file` (missed the plural, but I think it's still clear) in 6b71c300be3098da01db59e342088d090aa84cbb

File: src/condor_scripts/adstash/utils.py [hunk 3]
```diff
@@ -184,3 +185,22 @@ def collect_process_metadata():
...
+
+
+def parse_history_ad_file(adfile):
+    """
+    Generates one ClassAd at a time from adfile.
+    Necessary because classad.parseAds()
+    cannot handle files with "weird" ad separators
+    (e.g. "**** metadataA=foo metadata2=bar")
+    """
+    with open(adfile) as f:
```
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_scripts/adstash/utils.py]: Check added in 6b71c300be3098da01db59e342088d090aa84cbb

Example log error message:
`12/13/21 16:33:51 Could not read non-existing-file.ad: [Errno 2] No such file or directory: 'non-existing-file.ad'`

File: docs/man-pages/condor_adstash.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (markcoatsworth) [reply docs/man-pages/condor_adstash.rst]: LGTM

File: src/condor_scripts/adstash/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (markcoatsworth) [reply src/condor_scripts/adstash/utils.py]: LGTM

File: src/condor_scripts/adstash/history.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (markcoatsworth) [reply src/condor_scripts/adstash/history.py:45]: Sounds good.

**Review by markcoatsworth — APPROVED** [review]
All changes look good, thanks!

**Pull request merged**